### DISCUSSION TITLE
Why does deleteAll use two queries?

### DISCUSSION THREAD

**Discussion by timvandam** [prisma/prisma#8211] [Q&A] [answered]
I noticed that `deleteMany` uses two queries when I specify a `where`. It first selects the primary keys of the rows to delete, and then removes them with a `DELETE FROM WHERE id IN (...)` query.

What is the use of this? Instead of the `WHERE id IN (...)` query, it would make more sense to me to just select only the rows to delete in the `DELETE` query itself.

As an example:
```typescript
await prismaClient.cardSet.deleteMany({ where: { steamAccountId: steamAccount.id } });
```

This piece of code causes the first query to run, whose return values are then used in the second query:
```sql
SELECT "public"."CardSet"."id" FROM "public"."CardSet" WHERE "public"."CardSet"."steamAccountId" = $1;
DELETE FROM "public"."CardSet" WHERE "public"."CardSet"."id" IN ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,...);
```

The following seems more efficient to me:
```sql
DELETE FROM "public"."CardSet" WHERE "public"."CardSet"."steamAccountId" = $1;
```

Are there any cases where using just one simple query would not suffice?

**Comment by ryands17** [comment, marked as answer]
@timvandam 👋 

The reason we run two queries is because the first query was used to prevent cascade deletes. We used to guard against cascade deletes as we didn't support that yet: https://github.com/prisma/prisma/issues/6996

Now that we do support [cascade deletes](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/referential-actions) via a preview feature, we will be optimising this in the future 🙂

**Reply by timvandam** [reply]
Ah that makes sense. Thank you!

**Reply by ryands17** [reply]
Related request: https://github.com/prisma/prisma/issues/8239

**Reply by Jolg42** [reply]
Note: We did an optimization in `5.4.0`, see https://github.com/prisma/prisma-engines/pull/4249

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
